mmdet3d/models/detectors/h3dnet.py

```diff
@@ -0,0 +1,171 @@
+            gt_labels_3d, pts_semantic_mask, pts_instance_mask,
+            gt_bboxes_ignore)
+        losses.update(roi_losses)
+
+        return losses
+
+    def simple_test(self, points, img_metas, imgs=None, rescale=False):
+        """Forward of testing.
+
+        Args:
+            points (list[torch.Tensor]): Points of each sample.
+            img_metas (list): Image metas.
+            rescale (bool): Whether to rescale results.
+
+        Returns:
+            list: Predicted 3d boxes.
+        """
+        points_cat = torch.stack(points)
+
+        x = self.extract_feat(points_cat)
+        x['fp_xyz'] = [x['fp_xyz_net0'][-1]]
+        x['fp_features'] = [x['hd_feature']]
+        x['fp_indices'] = [x['fp_indices_net0'][-1]]
+
+        if self.with_rpn:
+            rpn_outs = self.rpn_head(x, self.test_cfg.rpn.sample_mod)
+            x.update(rpn_outs)
+            # Generate rpn proposals
+            rpn_proposals = self.rpn_head.get_bboxes(
+                points, rpn_outs, img_metas, with_nms=False)
+            x['rpn_proposals'] = rpn_proposals
+        else:
+            raise NotImplementedError
+
+        return self.roi_head.simple_test(
+            x,
+            self.test_cfg.rcnn.sample_mod,
+            img_metas,
+            points_cat,
+            rescale=rescale)
+
+    def aug_test(self, points, img_metas, imgs=None, rescale=False):
+        """Test with augmentation."""
+        points_cat = [torch.stack(pts) for pts in points]
+        feats = self.extract_feats(points_cat, img_metas)
+        for x in feats:
+            x['fp_xyz'] = [x['fp_xyz_net0'][-1]]
+            x['fp_features'] = [x['hd_feature']]
+            x['fp_indices'] = [x['fp_indices_net0'][-1]]
+
+        # only support aug_test for one sample
+        aug_bboxes = []
+        for x, pts_cat, img_meta in zip(feats, points_cat, img_metas):
+            if self.with_rpn:
+                rpn_outs = self.rpn_head(x, self.test_cfg.rpn.sample_mod)
+                x.update(rpn_outs)
+                # Generate rpn proposals
+                rpn_proposals = self.rpn_head.get_bboxes(
+                    points, rpn_outs, img_metas, with_nms=False)
+                x['rpn_proposals'] = rpn_proposals
+            else:
+                raise NotImplementedError
+
+            bbox_results = self.roi_head.simple_test(
+                x,
+                self.test_cfg.rcnn.sample_mod,
+                img_meta,
+                pts_cat,
+                rescale=rescale)
+            aug_bboxes.append(bbox_results)
+
+        # after merging, bboxes will be rescaled to the original image size
+        merged_bboxes = merge_aug_bboxes_3d(aug_bboxes, img_metas,
+                                            self.bbox_head.test_cfg)
+
+        return merged_bboxes
+
+    def extract_feats(self, points, img_metas):
+        """Extract features of multiple samples."""
+        return [
```

Comment:
1. docstrings
2. it seems like `points` is a Tensor, so is it necessary to use [... for ... in ...] ?